### PR TITLE
[CMS + KMS + Assume Role] Fix Bug in AmazonS3EncryptionClient#uploadO…

### PR DESCRIPTION
**Pull request by taklwu**
…bject by UploadObjectObserver#newInitiateMultipartUploadRequestquest

Fix bug when using uploadObject when there is RequestCredentialsProvider
set in the original upload request.

### CODE REVIEW

**Comment by taklwu** [comment]
it's related to https://github.com/aws/aws-sdk-java/issues/1851

mvn build passed, and tested the changes with our client program passed.

**Comment by taklwu** [comment]
ping ?

**Comment by taklwu** [comment]
although we can patched with our own release, it would be great if this code change is in mainline

File: aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java [hunk 1]
```diff
@@ -222,6 +223,7 @@ protected UploadPartRequest newUploadPartRequest(PartCreationEvent event,
             .withLastPart(event.isLastPart())
             .withUploadId(uploadId)
             .withObjectMetadata(req.getUploadPartMetadata())
+            .withRequestCredentialsProvider(req.getRequestCredentialsProvider())
```
Comment (zoewangg) [inline aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java]: Don't we need to add this in `s3#completeMultipartUpload` as well?

Comment (taklwu) [reply aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java]: sorry, can you be more specific on which class this `completeMultipartUpload ` function is being called ?

Comment (zoewangg) [reply aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java]: Yeah, sorry for being unclear. For Multipart upload request, we need to do three steps:

- Initiate multipartUpload request
- Upload each part
- Complete multipart upload

I noticed you added the requestCredentialProvider for the first two steps and I was wondering if we need to add it to the third step.

https://github.com/aws/aws-sdk-java/blob/0370178d6c0c0f40f6d0b8894e77ef8b133fd366/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java#L114

Comment (taklwu) [reply aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java]: So If I understand correctly `Complete multipart upload` would be called in the following lines in order

1. [observer.onCompletion(partETags);](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3EncryptionClient.java#L836) 
2. then, [UploadObjectObserver.java#onCompletion](https://github.com/aws/aws-sdk-java/blob/0370178d6c0c0f40f6d0b8894e77ef8b133fd366/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java#L179-L183)
3. [AmazonS3EncryptionClient.java#completeMultipartUpload](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3EncryptionClient.java#L597-L600)
4. [S3CryptoModuleBase.java#completeMultipartUploadSecurely()](https://github.com/aws/aws-sdk-java/blob/7bf0fbec42de8a7d54875510f27fb363ca2d19f5/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/internal/crypto/S3CryptoModuleBase.java#L377-L401)

above steps 3), 4) and `CompleteMultipartUploadRequest` does not have field to use/set `RequestCredentialsProvider` to their request, so and as far as our use cases that starts with `s3.uploadObject(req)` is using with `MultiFileOutputStream` within the request. 

So, I don't see the needs to change/add `RequestCredentialsProvider` to `Complete multipart upload`.

Comment (zoewangg) [reply aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java]: Yeah, `UploadObjectObserver.java#onCompletion` is what I was talking about. (I pasted a wrong link in my previous comment)

It creates a new `CompleteMultipartUploadRequest` request, so it seems odd to me that we only add `RequestCredentialsProvider` for other two requests and not `CompleteMultipartUploadRequest`. 

https://github.com/aws/aws-sdk-java/blob/0370178d6c0c0f40f6d0b8894e77ef8b133fd366/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java#L179-L183

Anyway, thanks for the information. We'll discuss internally of this change and provide an update shortly.

Comment (zoewangg) [reply aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java]: We discussed internally and we think we should add `RequestCredentialsProvider` to  `UploadObjectObserver#onCompletion` as well.

Could you confirm adding the requestCredentials to `UploadObjectObserver#onCompletion` would still work in your case?

Comment (taklwu) [reply aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java]: I see your point now, I will try to add `withRequestCredentialsProvider` in `CompleteMultipartUploadRequest` and its caller e.g. [`CompleteMultipartCopy.java#call()`](https://github.com/aws/aws-sdk-java/blob/7bf0fbec42de8a7d54875510f27fb363ca2d19f5/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CompleteMultipartCopy.java#L73-L103), [`UploadCallable.java#uploadPartsInSeries()`](https://github.com/aws/aws-sdk-java/blob/7bf0fbec42de8a7d54875510f27fb363ca2d19f5/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/UploadCallable.java#L242-L277) and [`CompleteMultipartUpload.java#call()`](https://github.com/aws/aws-sdk-java/blob/7bf0fbec42de8a7d54875510f27fb363ca2d19f5/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CompleteMultipartUpload.java#L84-L112).

I need a bit time to test each use case as well, as I mentioned our code only use `s3.uploadObject(req)` which won't cover all the corner cases. 

meanwhile, it's a bit out of my original scope to fix one problem, it's expended 4 use cases now, I'm wondered if we should on focus on my use case of `s3.uploadObject(req)` here...let me know if you or your team has any thoughts.

Comment (zoewangg) [reply aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/UploadObjectObserver.java]: Yeah, adding it to `UploadObjectObserver#onCompletion` like the code snippet you pasted above should be enough. Thanks!

**Comment by taklwu** [comment]
@zoewangg sorry for the late, updated `UploadObjectObserver#onCompletion` as suggested, `mvn` rebuilt passed, and tested with our environment and everything worked. please review it again, and let me know if you have any comments.

```
...
[INFO] AWS Java SDK for AWS Organizations ................. SUCCESS [  1.407 s]
[INFO] AWS Java SDK for Amazon WorkDocs ................... SUCCESS [  1.548 s]
[INFO] AWS SDK for Java - Models .......................... SUCCESS [ 13.127 s]
[INFO] AWS SDK For Java ................................... SUCCESS [  0.138 s]
[INFO] AWS SDK For Java for OSGi .......................... SUCCESS [01:30 min]
[INFO] AWS SDK for Java - OpenSdk Runtime ................. SUCCESS [  1.622 s]
[INFO] AWS SDK for Java - BOM ............................. SUCCESS [  0.006 s]
[INFO] AWS SDK for Java - Code Generator .................. SUCCESS [  2.107 s]
[INFO] AWS SDK for Java - Code Generator Maven Plugin ..... SUCCESS [  3.113 s]
[INFO] AWS SDK for Java - Bundle .......................... SUCCESS [ 35.071 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 08:34 min
[INFO] Finished at: 2019-02-12T09:39:40-08:00
[INFO] Final Memory: 301M/1803M
[INFO] ------------------------------------------------------------------------
```

**Comment by varunnvs92** [comment]
Changes lgtm. Thanks for the contribution. It will be included in the next release.
